### DISCUSSION TITLE
Multi Currency Feature

### DISCUSSION THREAD

**Discussion by josephelkhoury** [Blair2004/NexoPOS#1358] [Q&A]
Hello everyone,

I'm thinking of using NexoPOS for my business but a key feature is missing and I'm considering extending the POS with a custom module.

We need multi currency payment as in our country we have a variable local currency rate and it keeps changing.
So what I was thinking of was to create a module that lets us add new currencies and set the rate (that can be modified at any time).

On checkout, when the user will pick the payment method (cash, card...etc), he should also pick the currency and the total will be reflected depending on the rate.
And he can settle in several currencies: cash payment in usd, cash payment in other currency ...etc

Can this be done using modules only or I need to update the core code?
I know that I need to add a table for currencies and a column for the currency identifier in order_payments but will also need to add the currency to reporting, sales, cash registers (we'll need 1 cash register per currency) ...etc
A module implementation could do all that?
Should I edit the entities classes?

Your advice is highly appreciated.

Thanks.

**Comment by Blair2004** [comment]
Hi, 
Thank you for your consideration. A multi-currency feature is better implemented on the core. However, we don't want to add such feature at the moment on the core. It remains possible implementing on a module. While it's tricky, it remain possible. The solution will be to create a separate column for other device or you can convert in realtime (avoiding add many columns per currency).

If you still need to keep a reference, you can add a column that set the currency of the order. You need to be well organized to have that softly running.

Hope that helps.

**Reply by josephelkhoury** [reply]
I started working on it today as a module, hopefully it's gonna work.
But I have a question: Is it possible to extend an entity? I need to extend the OrderPayment entity to include the additional column.

**Reply by josephelkhoury** [reply]
Hi again,

I managed to create a module that does what I need.
But I have to make minor changes to some of the core files and I couldn't do it another way. 

I will tell you exactly what I added, maybe there is another way or a suggestion would be to add these minor changes in your next releases.

OrderAfterPaymentCreatedEvent.php
- Adding an additional parameter ($fields) to the constructor that will allow us to get the request data and make modifications to the OrderPayment object.

OrdersController.php
- Adding the new $payment parameter to the following event: 
`OrderAfterPaymentCreatedEvent::dispatch( $orderPayment, $order, $payment );`

pos-init.ts
- Add the following lines:
```
const nsCashPayment	=  (<any>window).nsCashPayment = require('./pages/dashboard/pos/payments/cash-payment').default;
const nsCreditCardPayment	=  (<any>window).nsCreditCardPayment = require('./pages/dashboard/pos/payments/creditcard-payment').default;
const nsBankPayment	=  (<any>window).nsBankPayment = require('./pages/dashboard/pos/payments/bank-payment').default;
const nsAccountPayment	=  (<any>window).nsAccountPayment = require('./pages/dashboard/pos/payments/account-payment').default;
const nsPosDiscountPopup = (<any>window).nsPosDiscountPopup = require('./popups/ns-pos-' + 'discount' + '-popup').default;
```

Let me know what you think.

**Reply by Blair2004** [reply]
Please make a pull request
